File: spring-boot-api/src/main/java/dev/codewithsam/springbootapi/payments/PaymentEntity.java
package dev.codewithsam.springbootapi.payments;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;
import org.hibernate.annotations.CreationTimestamp;

/**
 * The TypeORM equivalent is almost the same file, decorator for annotation.
 *
 * <p>The difference that does NOT show on the page: this object is MANAGED.
 * Load it inside a transaction, mutate a field, and Hibernate writes the change
 * back on flush without you calling save(). A TypeORM entity is a plain object
 * and nothing happens until you call save() yourself.
 */
@Entity
@Table(name = "payments")
public class PaymentEntity {

  @Id @GeneratedValue private UUID id;

  @Column(name = "amount_in_minor_units", nullable = false)
  private Long amountInMinorUnits;

  @Column(nullable = false, length = 3)
  private String currency;

  @Column(name = "idempotency_key", nullable = false, length = 100)
  private String idempotencyKey;

  @CreationTimestamp
  @Column(name = "created_at", nullable = false, updatable = false)
  private Instant createdAt;

  protected PaymentEntity() {}

  public PaymentEntity(Long amountInMinorUnits, String currency, String idempotencyKey) {
    this.amountInMinorUnits = amountInMinorUnits;
    this.currency = currency;
    this.idempotencyKey = idempotencyKey;
  }

  public UUID getId() {
    return id;
  }

  public Long getAmountInMinorUnits() {
    return amountInMinorUnits;
  }

  public String getCurrency() {
    return currency;
  }

  public String getIdempotencyKey() {
    return idempotencyKey;
  }

  public Instant getCreatedAt() {
    return createdAt;
  }

  public Payment toPayment() {
    return new Payment(
        "pay_" + id, amountInMinorUnits, currency, idempotencyKey, createdAt);
  }
}

File: nestjs-api/src/payments/payment.entity.ts
import {
  Column,
  CreateDateColumn,
  Entity,
  PrimaryGeneratedColumn,
  Unique,
} from 'typeorm';

// SIDE BY SIDE WITH THE JPA ENTITY THIS IS ALMOST THE SAME FILE.
//   @Entity            -> @Entity()
//   @Id @GeneratedValue -> @PrimaryGeneratedColumn('uuid')
//   @Column            -> @Column()
//   @CreationTimestamp -> @CreateDateColumn()
//
// The resemblance is exactly why this is dangerous. A JPA entity is MANAGED:
// load it, mutate a field, and Hibernate writes the change back on flush
// without you calling anything. A TypeORM entity is a plain object. Mutate it
// and nothing happens until you call save(). That difference is measured in
// the persistence episode rather than asserted here.
// THE UNIQUE CONSTRAINT IS THE THING THAT ACTUALLY CLOSES THE RACE.
// A transaction narrows the window between the lookup and the insert. It does
// not remove it, because two transactions can both look, both find nothing,
// and both insert. Only the database can refuse the second one.
@Entity('payments')
@Unique('uq_payments_idempotency_key', ['idempotencyKey'])
export class PaymentEntity {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  // bigint maps to a STRING in the JavaScript driver, not a number, because a
  // 64-bit integer does not fit in a JavaScript number without losing
  // precision. This is the single most surprising line in the file for a Java
  // developer and it gets its own beat.
  @Column({ type: 'bigint' })
  amountInMinorUnits!: string;

  @Column({ type: 'varchar', length: 3 })
  currency!: string;

  @Column({ type: 'varchar', length: 100, name: 'idempotency_key' })
  idempotencyKey!: string;

  @CreateDateColumn({ name: 'created_at' })
  createdAt!: Date;
}
